calyx-backend/src/xilinx/idl.rs

```diff
@@ -0,0 +1,148 @@
+use crate::traits::Backend;
+use calyx_ir as ir;
+use calyx_utils::CalyxResult;
+use serde::Serialize;
+/// Backend that generates the Interface Design Language
+/// that we need to create AXI wrappers on arbitrary programs
+#[derive(Default)]
+pub struct IdlBackend;
+
+/// The root element of the `kernel.xml` file that describes an `.xo` package for the
+/// Xilinx toolchain, as documented [in the Vitis user guide][ug].
+///
+/// [ug]: https://docs.xilinx.com/r/en-US/ug1393-vitis-application-acceleration/RTL-Kernel-XML-File
+#[derive(Serialize)]
+struct ProgramInterface<'a> {
+    toplevel_name: &'a str,
+    memories: Vec<Memory<'a>>,
+}
+
+#[derive(Serialize)]
+struct Memory<'a> {
+    name: &'a str,
+    width: u64,
+    size: u64, //size of width size memory, as defined in stdlib.
+}
+
+impl Backend for IdlBackend {
+    fn name(&self) -> &'static str {
+        "idl"
+    }
+
+    fn validate(_ctx: &ir::Context) -> CalyxResult<()> {
+        Ok(())
+    }
+
+    fn link_externs(
+        _prog: &ir::Context,
+        _write: &mut calyx_utils::OutputFile,
+    ) -> CalyxResult<()> {
+        Ok(())
+    }
+
+    fn emit(
+        prog: &ir::Context,
+        file: &mut calyx_utils::OutputFile,
+    ) -> CalyxResult<()> {
+        let toplevel = prog
+            .components
+            .iter()
+            .find(|comp| comp.name == prog.entrypoint)
+            .unwrap();
+
+        let memory_names = external_memories(toplevel);
+        let mem_infos = get_mem_info(toplevel);
+
+        let memories: Vec<Memory> = memory_names
+            .iter()
+            .zip(mem_infos.iter())
+            .map(|(memory_name, mem_info)| Memory {
+                name: memory_name,
+                width: mem_info.width,
+                size: mem_info.size,
+            })
+            .collect();
+
+        let program_interface = ProgramInterface {
+            toplevel_name: toplevel.name.as_ref(),
+            memories,
+        };
+
+        //TODO: we want to have na array of component names
+        // check how to get them
+
+        write!(
+            file.get_write(),
+            "{}",
+            serde_json::to_string_pretty(&program_interface)
+                .expect("IDL Serialization failed")
+        )?;
```

Comment:
I could be wrong, but I think there is a slightly simpler way to write to the file directly without making a string first?
https://docs.rs/serde_json/latest/serde_json/fn.to_writer.html

As in, something like:

```rs
serde_json::to_writer(file.get_write(), &program_interface);
```